### ISSUE TITLE
Should we define multiple proof / path representations?

### ISSUE THREAD

**Issue by OR13** [ietf-scitt/cose-merkle-tree-proofs#6]
Originally: 

TODO: Do we need both inclusion path types? what properties does each type have?

### Option 0

```
IndexAwareInclusionPath = [
    leaf_index: int
    hashes: [+ bstr]
]
```

### Option 1

```
IndexUnawareInclusionPath = [+ PathEntry]
PathEntry = [
    left: bool
    hash: bstr
]
```

**Comment by OR13** [comment]
Are both options equally performant in balanced and unbalanced trees?

Do both options reveal the same information to a verifier?

Does index positioning reveal "order over time time" ?

**Comment by OR13** [comment]
There is some bias towards option 0 if we believe that communicating tree size happens in the header.

**Comment by OR13** [comment]
I'm not a "COSE / CBOR" encoding expert, but it seems that there is a more compact representation for "index unaware proofs"

```
IndexUnawareInclusionPath = [
    hashes: [+ [ bool, bstr ] ]
]
```

For large tree sizes where the size is not revealed it is more compact, since 2 int can be substituted for 1 bit per hash.

**Comment by letmaik** [comment]
There's also the variant that QLDB uses where the node hashes are sorted before hashing such that at inclusion path verification you don't need the direction or leaf index because you repeat the same sorting and know implicitly whether it goes left or right: see also [OpenZeppelin](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/8177c4620e049b2749c2069651d7d5b4691e23d2/contracts/utils/cryptography/MerkleProof.sol#L211-L213). Obviously this only works for that specific tree type.

**Comment by OR13** [comment]
Another consideration is "multiple proof disclosures"...

Its possible that one option performs better than the other after a certain percentage of the tree is revealed.

**Comment by OR13** [comment]
As a principle, it would be nice to define the "most compact possible" representation... even if its not the "most popular one".

**Comment by OR13** [comment]
@letmaik yes that works as well... in my first attempt to make compact proofs I used a bit string for directions, and then concatenated with hashes, it looked sorta like this:

`010... hash-0, hash-1, ...`

https://github.com/transmute-industries/verifiable-data/blob/1319ff2bf72ec7da1a90acfee7a424ab2962ef76/packages/merkle-proof/src/index.ts#L203

I removed this (and the pako compression) in the next revision to keep the algorithms simpler / cut down on dependencies.

**Comment by OR13** [comment]
Example QLDB Proof with other stuff wrapped around it:

```json
{
  "Proof": {
    "IonText": "[{{xwtiJrjaFEezJtfPnZZJkZniHpif/Ko58kEcdrSvUD0=}},{{DkEJ8WxOwHcrLtvv+hCmur3LIuXm5TP4qI+oFdV+nPI=}},{{H5SvpW4VsFRBbl2XEpJpAydwetVSixUi+epX7CD8FWY=}},{{UWx7JmpQLZEDdprLGz22cfBbjZkwFKIDjOj5ejRueVQ=}},{{4tujyvTE8a0N3pSk8ItGjkxho7mtOyvIm/mSW9LtrA4=}},{{eIWAJefiyGtwquKvX3dxs50g65aWiC8457IR1CAG8wg=}},{{BCAa4xj0NhToMx694Tx+d7dlRpF8cWIApDLVzgu26pY=}},{{5UgPeK4mx6ryvdkBZwkfOV1lsR60+7WLpRFKCj/DGX8=}},{{PlGnVVc976YN5ven3wluckeFTpEpzbC6Xicjg0+51aU=}},{{NQWgMUnpqoKrpyTsZgkasf0NxHJe1Wi5LdmpfNFNin0=}},{{YGmPTziy2xt7wU6n3QaFCbBKE7FMr82CajoMz/PQX7s=}},{{OjELd9pepTZiojuX3RkEbvo6UdBlPM3m+ht0RVSmobY=}},{{8AqUg8lzSIiAfUVWmtFjBKQGFK2gHwAOE9r/G403YA4=}}]"
  },
  "Revision": {
    "IonText": "{blockAddress:{strandId:\"5EhcZ4QwhGvDaV3PN3GoRd\",sequenceNo:672},hash:{{k07YWpQB13KYIKWovxa3GPuc+b+WmoWA3rUK7Dx1Mz8=}},data:{lastName:\"Doe\",age:42,firstName:\"Timmy\"},metadata:{id:\"7j4KlV1swnTBGJhuF6l4kZ\",version:0,txTime:2023-02-26T16:33:34.423Z,txId:\"47VIqjNef3N9p85eEn0dDO\"}}"
  }
}
```

**Comment by letmaik** [comment]
Another variant of inclusion proof is simply providing all other members by hash. Then the whole tree has to be re-built to get to the root. This would only realistically work for the disclosure case (as the trees are typically small), rather than transparency logs. I wonder if there's ever a point to choose this form compared to path-based proofs.

**Comment by OR13** [comment]
@fournet and I discussed, we feel that each tree alg should define only 1 inclusion path representation, and supporting multiple inclusion path representations for the same tree alg should be discouraged.

**Comment by OR13** [comment]
We propose to close this issue in https://github.com/ietf-scitt/draft-steele-cose-merkle-tree-proofs/pull/9
